### PR TITLE
Mode, median and stdev in two articles

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+---
+title: "Calculate the mode & median averages of a Ruby array"
+description: "A lesson in why we use native Ruby methods"
```
Comment (andycroll) [inline _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]:
```suggestion
description: "An average isn't just a mean average."
```

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+end
+#=> 3
+```
+
+
+## Why?
+
+The `#tally` method was added to `Enumerable` in Ruby 2.7 so you might see implementations that use `inject` which will perform worse.
+
+I compared the performance of the different implementations when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array).
```
Comment (andycroll) [inline _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]:
```suggestion
I compared the performance of using native Ruby methods versus your own implementations when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array).
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+
+## Why?
+
+Using [the `#sum` method from `Array`](https://ruby-doc.org/core-3.0.0/Array.html#method-i-sum) is many, many times faster than using the alternative, `inject`.
+
+The `#sum` method was added to `Array` in Ruby 2.4, which is why you might see alternative implementations in other places on the Internet.
+
+I compared the performance of the different implementations when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array).
```
Comment (andycroll) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
I compared the performance of Ruby-native vs. implementing the algorithms yourself when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array) and the same principles apply: that native implementations (in C) are much faster.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+
+## Anything else?
+
+In all honesty, if you’re doing a lot of statistical number-crunching work you probably want to reach a little closer to “the metal”.
+
+A version of the standard deviation calculation done in Ruby is much slower than if it were done natively in C. This 'lower-level' support is why running `sum` is much faster than running `inject` in the above example, because Ruby’s `sum` method is implemented in C.
+
+If you're doing a lot of this sort of calculation or in a situation performance is key you might want to look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented (in C, that are called from Ruby) versions of several statistical summary methods mixed in directly to `Array`s and `Enumerable`s.
```
Comment (andycroll) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
If you're doing a lot of this sort of calculation or in a situation performance is key you might want to look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented (in C, called from Ruby) versions of several statistical summary methods mixed in directly to Ruby's `Array` and `Enumerable` classes.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+I compared the performance of Ruby-native vs. implementing the algorithms yourself when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array) and the same principles apply: that native implementations (in C) are much faster.
+
+
+## Anything else?
+
+In all honesty, if you’re doing a lot of statistical number-crunching work you probably want to reach a little closer to “the metal”.
+
+A version of the standard deviation calculation done in Ruby is much slower than if it were done natively in C. This 'lower-level' support is why running `sum` is much faster than running `inject` in the above example, because Ruby’s `sum` method is implemented in C.
+
+If you're doing a lot of this sort of calculation or in a situation performance is key you might want to look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented (in C, that are called from Ruby) versions of several statistical summary methods mixed in directly to `Array`s and `Enumerable`s.
```
Comment (andycroll) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
If you're doing a lot of this sort of calculation or in a situation performance is key you might want to look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented (in C, called from Ruby) versions of several statistical summary methods mixed in directly to Ruby's `Array` and `Enumerable` classes.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+standard_deviation = Math.sqrt(variance)
+#=> 2.449489742783178
+```
+
+
+## Why?
+
+Using [the `#sum` method from `Array`](https://ruby-doc.org/core-3.0.0/Array.html#method-i-sum) is many, many times faster than using the alternative, `inject`.
+
+The `#sum` method was added to `Array` in Ruby 2.4, which is why you might see alternative implementations in other places on the Internet.
```
Comment (andycroll) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
The `#sum` method was only added to `Array` in Ruby 2.4, which is why you might see alternative implementations in other places on the Internet.
```

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+category: ruby
+image:
+  base: '2020/calculate-a-mean-average-from-a-ruby-array'
+  alt: "Black and White Texas Instruments calculator"
+  credit:  Ray Reyes
+  source: "https://unsplash.com/photos/3xwrg7Vv6Ts"
+
+---
+
+I've previously disscussed a way [to calculate the mean of an array in Ruby](/ruby/calculate-a-mean-average-from-a-ruby-array). But there are two other averages that you might see, less commonly used but still useful.
```
Comment (nodunayo) [inline _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]:
```suggestion
I've previously disscussed a way [to calculate the mean of an array in Ruby](/ruby/calculate-a-mean-average-from-a-ruby-array). But there are two other averages that you might see.. They are less frequently used, but still powerful.
```

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+else
+  top_pair[1][0]
+end
+#=> 3
+```
+
+
+## Why?
+
+The `#tally` method was added to `Enumerable` in Ruby 2.7 so you might see implementations that use `inject` which will perform worse.
```
Comment (nodunayo) [inline _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]: * comma before 'which'

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+I compared the performance of using native Ruby methods versus your own implementations when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array).
+
+
+## Anything else?
+
+It's no doubt a good idea to further encapsulate these calulations into methods.
+
+However you're doing a lot of this sort of calculation or in a situation performance is important you should look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented (in C, that are called from Ruby) versions of several statistical summary methods mixed in directly to `Array`s and `Enumerable`s, and thus is much faster than any version of the algorithm in Ruby.
```
Comment (nodunayo) [inline _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]: *'if' missing?

*situation _where_ performance is important...

*the bit in parentheses doesn't read so well

* Array

* Enumerable

* and is _therefore_ much faster

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md (hunk 7 above)
Comment (andycroll) [inline _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]:
```suggestion
I've previously disscussed a way [to calculate the mean of an array in Ruby](/ruby/calculate-a-mean-average-from-a-ruby-array). But there are two other averages that you might see, less commonly used, but still useful.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+category: ruby
+image:
+  base: '2020/calculate-a-mean-average-from-a-ruby-array'
+  alt: "Black and White Texas Instruments calculator"
+  credit:  Ray Reyes
+  source: "https://unsplash.com/photos/3xwrg7Vv6Ts"
+
+---
+
+Standard deviation is a measure of the amount of variance of group of values. A high value indicates a wide spread of values around the mean, a low value indicates tight clustering of values.
```
Comment (nodunayo) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]: *within/amongst a group of values?

*,_whereas_ a low value indicates a tight clustering.

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+The calculation is used to understand and give context to the mean value of a series of data, [wikipedia has some good examples of real-world applications](https://en.wikipedia.org/wiki/Standard_deviation#Application_examples).
```
Comment (nodunayo) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]: Full stop before 'Wikipedia'.

I dunno: "understand and give context too..." sounds wishy washy.

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+Ruby doesn’t provide a native method to generate the standard deviation of an array of integers. Its built-in [`Math`](https://ruby-doc.org/core-3.0.0/Math.html) library focuses on trigonometry and logarithmic calculations. Given there is no built-in way [to calculate the mean of an array in Ruby](/ruby/calculate-a-mean-average-from-a-ruby-array) there’s equally no way to calculate the standard deviation.
```
Comment (nodunayo) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]: Sounds like you're saying the same thing twice...! I know it's slightly different, but it reads rather clunky.

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+
+## Why?
+
+Using [the `#sum` method from `Array`](https://ruby-doc.org/core-3.0.0/Array.html#method-i-sum) is many, many times faster than using the alternative, `inject`.
+
+The `#sum` method was only added to `Array` in Ruby 2.4, which is why you might see alternative implementations in other places on the Internet.
+
+I compared the performance of Ruby-native vs. implementing the algorithms yourself when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array) and the same principles apply: that native implementations (in C) are much faster.
```
Comment (nodunayo) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]: *drop the 'that' after the colon and before 'native'

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+
+## Anything else?
+
+In all honesty, if you’re doing a lot of statistical number-crunching work you probably want to reach a little closer to “the metal”.
+
+A version of the standard deviation calculation done in Ruby is much slower than if it were done natively in C. This 'lower-level' support is why running `sum` is much faster than running `inject` in the above example, because Ruby’s `sum` method is implemented in C.
```
Comment (nodunayo) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]: Again, repetitive. I say drop everything after 'above example'. You've already said it's native in C.

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+If you're doing a lot of this sort of calculation or in a situation performance is key you might want to look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented (in C, called from Ruby) versions of several statistical summary methods mixed in directly to Ruby's `Array` and `Enumerable` classes.
```
Comment (nodunayo) [inline _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]: *_where_ performance is key

Again "in C, called from Ruby" reads a bit clunkily.

**Review by nodunayo — CHANGES_REQUESTED** [review]

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md (hunk 8 above)
Comment (andycroll) [reply _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]:
```suggestion
The `#tally` method was added to `Enumerable` in Ruby 2.7 so you might see implementations that use `inject`, which will perform worse.
```

File: _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md (hunk 9 above)
Comment (andycroll) [reply _posts/2021/2021-06-21-calculate-mode-median-average-of-a-ruby-array.md]:
```suggestion
If you're doing a lot of this sort of calculation or in a situation where performance is important you should look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented versions of several statistical summary methods mixed into `Array` and `Enumerable`.

The implementation in C means these methods are much faster than any version of the algorithm programmed in Ruby.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md (hunk 10 above)
Comment (andycroll) [reply _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
Standard deviation is a measure of the amount of variance within a group of values. A high value indicates a wide spread of values around the mean, whereas a low value indicates tight clustering of values.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md (hunk 11 above)
Comment (andycroll) [reply _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
The calculation gives additional context to the range of values around the mean value of a series of data. [Wikipedia has some good examples of real-world applications](https://en.wikipedia.org/wiki/Standard_deviation#Application_examples).
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md (hunk 12 above)
Comment (andycroll) [reply _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
Ruby doesn’t provide a native method to generate the standard deviation of an array of integers. Its built-in [`Math`](https://ruby-doc.org/core-3.0.0/Math.html) library focuses on trigonometry and logarithmic calculations. This isn't surprising given there's not even a way [to calculate the mean of an array in Ruby](/ruby/calculate-a-mean-average-from-a-ruby-array).
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md (hunk 13 above)
Comment (andycroll) [reply _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
I compared the performance of Ruby-native vs. implementing the algorithms yourself when I wrote about [calculating the mean](/ruby/calculate-a-mean-average-from-a-ruby-array) and the same principles apply: native implementations (in C) are much faster.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md (hunk 14 above)
Comment (andycroll) [reply _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
A version of the standard deviation calculation done in Ruby is much slower than if it were done natively in C.
```

File: _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md (hunk 15 above)
Comment (andycroll) [reply _posts/2021/2021-06-28-calculate-the-standard-deviation-of-a-ruby-array.md]:
```suggestion
If you're doing a lot of this sort of calculation or in a situation where performance is key you might want to look at the [`enumerable-statistics`](https://github.com/mrkn/enumerable-statistics) gem. It has natively implemented versions of several statistical summary methods mixed in directly to Ruby's `Array` and `Enumerable` classes.
```

**Pull request merged**